core/dataset.cpp

```diff
@@ -758,4 +814,38 @@ std::unordered_map<Dim, scipp::index> Dataset::dimensions() const {
   return all;
 }
 
+std::map<typename MasksConstProxy::key_type,
+         typename MasksConstProxy::mapped_type>
+union_or(const MasksConstProxy &currentMasks,
+         const MasksConstProxy &otherMasks) {
+  std::map<typename MasksConstProxy::key_type,
+           typename MasksConstProxy::mapped_type>
+      out;
+
+  for (const auto & [ key, item ] : currentMasks) {
+    out.emplace(key, item);
+  }
+
+  for (const auto & [ key, item ] : otherMasks) {
+    const auto it = currentMasks.find(key);
+    if (it != currentMasks.end()) {
+      out[key] |= item;
+    } else {
+      out.emplace(key, item);
+    }
+  }
+  return out;
+}
+
+void union_or(const MasksProxy &currentMasks,
```

Comment:
Can you rename this function? You risk calling the wrong overload, in particular since `MasksConstProxy` is a base class of `MasksProxy`.

Reply:
`union_or_in_place`? Not a good name, but I would not worry too much given that it is internal-use only.